Fixed card duplication issue #2349

matuspetro opened the pull request:
1. Removed condition that blocked image comparison
2. Changed comparison logic from ID-based to cardID-based to prevent duplicates on repeated import

matuspetro commented:
@TheLastProject

TheLastProject commented on app/src/main/java/protect/card_locker/DBHelper.java:
```diff
@@ -561,6 +561,30 @@ public static LoyaltyCard getLoyaltyCard(Context context, SQLiteDatabase databas
+        data.close();
+        return cards;
+    }
+
```
What is this function for? Why would you need to get a list of the cards by what's in the barcode? I think you're probably confusing "id" with "cardId" ("id" being the unique identifier").

You probably want to just use `getLoyaltyCard`

matuspetro replied:
Yes, I know it's a barcode. I didn't want to compare the imported card with all the others, so I chose the barcode as a filter.

Previously, the code only compared the card with an existing card with the same ID, which led to duplicates. Since the ID is assigned randomly, it must be removed from the comparison.

This part of the code is therefore essential from a performance perspective.

TheLastProject replied:
> This part of the code is therefore essential from a performance perspective.

I mean, not really, you can also just call one of the existing `getLoyaltyCardCursor` functions and use the store name as the filter. In fact that would be better, because a cursor lets you loop over the results without loading them all into memory :)

I do agree that removing the database ID from the comparison is correct though.

TheLastProject commented on app/src/main/java/protect/card_locker/importexport/CatimaImporter.java:
```diff
@@ -147,51 +122,72 @@ public Map<Integer, Integer> saveAndDeduplicate(Context context, SQLiteDatabase
         return true;
     }
 
+    private long insertCard(Context context, SQLiteDatabase database, LoyaltyCard card, boolean newCard) throws IOException{
```
Code style

```suggestion
    private long insertCard(Context context, SQLiteDatabase database, LoyaltyCard card, boolean newCard) throws IOException {
```

TheLastProject commented on app/src/main/java/protect/card_locker/importexport/CatimaImporter.java:
```diff
@@ -147,51 +122,72 @@ public Map<Integer, Integer> saveAndDeduplicate(Context context, SQLiteDatabase
+        }else {
+            DBHelper.insertLoyaltyCard(database, card.id, card.store, card.note, card.validFrom, card.expiry, card.balance, card.balanceType,
+                    card.cardId, card.barcodeId, card.barcodeType, card.headerColor, card.starStatus, card.lastUsed, card.archiveStatus);
+        }
```
Given you only insert cards now if they don't match an existing card, isn't it always with `newCard` as `True`? I don't think you need this if/else anymore?

matuspetro replied:
This was just so that I wouldn't call the same thing four times. No, newCard is not always true. But I remove this in next commit.

TheLastProject replied:
I don't understand why newCard will not always be true.

Either it's a duplicate, so you don't insert it.
Or it's not a duplicate, so it gets inserted as a new card.

matuspetro commented on app/src/main/java/protect/card_locker/importexport/CatimaImporter.java:
```diff
@@ -147,51 +122,72 @@ public Map<Integer, Integer> saveAndDeduplicate(Context context, SQLiteDatabase
+        Utils.saveCardImage(context, card.getImageFront(context), card.id, ImageLocationType.front);
+        Utils.saveCardImage(context, card.getImageBack(context), card.id, ImageLocationType.back);
+
+        return newId;
```
As I wrote above, this code will not be in the next commit.

TheLastProject commented on app/src/main/java/protect/card_locker/importexport/CatimaImporter.java:
```diff
@@ -59,79 +58,55 @@ public static class ImportedData {
+                }else{
+                    insertCard(context, database, card, false);
+                }
             }
```
Sounds more like the tests are wrong in this case then? Because that doesn't sound very logical :sweat_smile:

TheLastProject reviewed (changes requested):
I still need to actually test this code again (and I barely have time to look really deeply into it), but two issues remain unsolved.

I think this is very close to being merge-worthy though

TheLastProject commented:
Don't know if this is for Hacktoberfest or not. I don't want you to get "screwed over" for this not being completed in time, so I'll label it hacktoberfest-accepted just in case :) Would be great if the last comments can be looked at, I think this is probably almost correct, but it's clearly already deserved to count from the effort put into it :)

matuspetro commented:
@TheLastProject I removed additional function getLoyaltyCardsByCardId and I call instead DB cursor as you wanted.

I look at the second bug too. When I removed checking of existing card with the same ID, it created 3 errors in test as last time, but I was not able to fix them. I find out that the problem was in importing cards to group, but I didn't see where. It added card to group, but the card wasn't existing. So number cards in group was higher that in reality. It little bit surprised me, how it's possible that problem is not there when I add this checking. In conclusion, it is good as it is.

I have another work too, so in my point of view this issue is done. If you test it, and you will find any duplication of cards or data lost, please let me know and I will fix it. But if this code will work fine, I will be waiting for merge.

I'm not involved in Hacktoberfest, but I will be very thankful if you will merge this 😊
